### PR TITLE
Add code eval

### PR DESCRIPTION
**Pull request by samhavens**
This is a re-do of Rishab's PR #441 , but now we have a stable version of Composer.

This has been tested somewhat but there are issues with nondeterminism. As well, we are still figuring out how to eval using the secure sandbox on AWS.

### CODE REVIEW

File: scripts/train/yamls/finetune/mpt-7b-code.yaml
Comment (bmosaicml) [inline scripts/train/yamls/finetune/mpt-7b-code.yaml:1]: Does this file need to be checked in?

File: scripts/eval/yamls/tasks.yaml [hunk 1]
```diff
@@ -173,3 +173,27 @@ icl_tasks:
   num_fewshot: [10]
   icl_task_type: multiple_choice
   continuation_delimiter: "\nAnswer: " # this separates questions from answers
+-
+  label: human_eval
+  dataset_uri: eval/local_data/programming/human_eval.jsonl # ADD YOUR OWN DATASET URI
+  num_fewshot: [0]
```
Comment (bmosaicml) [inline scripts/eval/yamls/tasks.yaml:179]: Do you mind making these all 3-shot?

File: scripts/eval/yamls/eval_gauntlet.yaml [hunk 2]
```diff
@@ -112,6 +112,17 @@ eval_gauntlet:
     - name: boolq
       num_fewshot: 10
       random_baseline: 0.5
+  - name: programming
+    benchmarks:
+    - name: human_eval
+      num_fewshot: 0
```
Comment (bmosaicml) [inline scripts/eval/yamls/eval_gauntlet.yaml:118]: 3 shot needed here as well

**Review by bmosaicml — APPROVED** [review]
Lgtm, just left some comments

File: scripts/eval/yamls/tasks.yaml (hunk 1 above)
Comment (bmosaicml) [reply scripts/eval/yamls/tasks.yaml:179]: let's do both zero and three shot: [0,3]

**Comment by samhavens** [comment]
I used the following YAML to run the new coding evals against MPT-7b-8k. I reproduced our previous value for HumanEval (Python) of 11.6%

```yaml
integrations:
- integration_type: git_repo
  git_repo: mosaicml/llm-foundry
  git_branch: sam/add-coding-eval
  pip_install: -e ".[gpu]"
  ssh_clone: false
- integration_type: wandb
  project: code-lora
  tags:
  - peft
  - eval
  entity: mosaic-ml

command: |
  cd llm-foundry/scripts
  composer eval/eval.py /mnt/config/parameters.yaml  || (echo "Command failed - killing python" && pkill python && exit 1)


run_name: humaneval-mpt-7b-8k
gpu_num: 8
gpu_type: a100_40gb
cluster: r7z2

resumable: false
priority: medium

image: mosaicml/examples:llm-latest

env_variables:
  - key: DATABRICKS_HOST
    value: redacted
  - key: DATABRICKS_TOKEN
    value: redacted
  - key: CODE_EVAL_DEVICE
    value: LOCAL

parameters:
  dist_timeout: 6000
  seed: 1
  max_seq_len: 1024
  device_eval_batch_size: 4
  precision: amp_fp16

  loggers:
    wandb: {}
    mlflow:
      experiment_name: redacted
      tracking_uri: databricks

  model_name_or_path: mosaicml/mpt-7b-8k

  models:
  -
    model_name: ${model_name_or_path}
    model:
      name: hf_causal_lm
      pretrained_model_name_or_path: ${model_name_or_path}
      init_device: cpu
      pretrained: true
    tokenizer:
      name: ${model_name_or_path}
      kwargs:
        model_max_length: ${max_seq_len}

  fsdp_config:
    sharding_strategy: FULL_SHARD
    mixed_precision: PURE
    activation_checkpointing: true
    activation_checkpointing_reentrant: false
    activation_cpu_offload: false
    limit_all_gathers: true
    verbose: false

  icl_tasks: 'eval/yamls/coding_tasks.yaml'
```

Note that this runs the code eval "locally" (on the same machine as eval.py). We need to very this works in Lambdas as well.

**Comment by mcarbin** [comment]
Repro using @samhavens yaml produces different results:

| Category  | Benchmark   | Subtask  |  Accuracy | Number few shot  | Model       |
|:-----------|:---------------|:----------|-----------:|:------------------|:-------------------|
|      | human_eval   |      | 0.0914634 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_cpp |      | 0.0372671 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_js |      | 0.0426829 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_c  |      | 0.111111 | 0-shot      | mosaicml/mpt-7b-8k |


@samhavens and I suspected generate/decode params and @samhavens found sampling is enabled for code eval, hence introducing nondeterminism: 

https://github.com/mosaicml/composer/blob/71cdfad53c5962fed8bcf340bb2bbfe403472678/composer/datasets/in_context_learning_evaluation.py#L1048

**Comment by dakinggg** [comment]
@samhavens @mcarbin I would still expect it to be deterministic if we are seeding everything properly...if you run on the same hardaware, batch size, etc, the data order should be the same. Perhaps we should see closer to the generate call to make it more deterministic.

**Comment by mcarbin** [comment]
non-determinism due to a subtle bug in seeding eval trainer. fixed with  [a2c877b](https://github.com/mosaicml/llm-foundry/pull/587/commits/a2c877b10f1e93f17f42f710b63cf2cf5b548fd9)

# before

| Category  | Benchmark   | Subtask  |  Accuracy | Number few shot  | Model       |
|:-----------|:---------------|:----------|-----------:|:------------------|:-------------------|
|      | human_eval   |      | 0.0914634 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_cpp |      | 0.0372671 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_js |      | 0.0487805 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_c  |      | 0     | 0-shot      | mosaicml/mpt-7b-8k |

| Category  | Benchmark   | Subtask  |  Accuracy | Number few shot  | Model       |
|:-----------|:---------------|:----------|-----------:|:------------------|:-------------------|
|      | human_eval   |      | 0.0731707 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_cpp |      | 0.0496894 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_js |      | 0.0243902 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_c  |      | 0     | 0-shot      | mosaicml/mpt-7b-8k |

| Category  | Benchmark   | Subtask  |  Accuracy | Number few shot  | Model       |
|:-----------|:---------------|:----------|-----------:|:------------------|:-------------------|
|      | human_eval   |      | 0.109756 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_cpp |      | 0.0248447 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_js |      | 0.0426829 | 0-shot      | mosaicml/mpt-7b-8k |
|      | human_eval_c  |      | 0     | 0-shot      | mosaicml/mpt-7b-8k |

# after 


| Category   | Benchmark      | Subtask   |   Accuracy | Number few shot   | Model              |
|:-----------|:---------------|:----------|-----------:|:------------------|:-------------------|
|            | human_eval     |           |  0.109756  | 0-shot            | mosaicml/mpt-7b-8k |
|            | human_eval_cpp |           |  0.0372671 | 0-shot            | mosaicml/mpt-7b-8k |
|            | human_eval_js  |           |  0.0365854 | 0-shot            | mosaicml/mpt-7b-8k |
|            | human_eval_c   |           |  0         | 0-shot            | mosaicml/mpt-7b-8k |


| Category   | Benchmark      | Subtask   |   Accuracy | Number few shot   | Model              |
|:-----------|:---------------|:----------|-----------:|:------------------|:-------------------|
|            | human_eval     |           |  0.109756  | 0-shot            | mosaicml/mpt-7b-8k |
|            | human_eval_cpp |           |  0.0372671 | 0-shot            | mosaicml/mpt-7b-8k |
|            | human_eval_js  |           |  0.0365854 | 0-shot            | mosaicml/mpt-7b-8k |
|            | human_eval_c   |           |  0         | 0-shot            | mosaicml/mpt-7b-8k |



| Category   | Benchmark      | Subtask   |   Accuracy | Number few shot   | Model              |
|:-----------|:---------------|:----------|-----------:|:------------------|:-------------------|
|            | human_eval     |           |  0.109756  | 0-shot            | mosaicml/mpt-7b-8k |
|            | human_eval_cpp |           |  0.0372671 | 0-shot            | mosaicml/mpt-7b-8k |
|            | human_eval_js  |           |  0.0365854 | 0-shot            | mosaicml/mpt-7b-8k |
|            | human_eval_c   |           |  0         | 0-shot            | mosaicml/mpt-7b-8k |

**Comment by mcarbin** [comment]
Non-determinism also exposed an issue with code eval pass@k metric. fixing is WIP

**Comment by dakinggg** [comment]
Closes #362

**Comment by mcarbin** [comment]
Waiting on PR to composer: https://github.com/mosaicml/composer/pull/2550

**Review by dakinggg — COMMENTED** [review]
@bmosaicml could you rereview please? want to make sure the gauntlet changes are ok with you

**Comment by dakinggg** [comment]
Also, please bump the `msoaicml` version pin in `setup.py` to `>=0.16.3`. Then once its released CI should pass

**Review by dakinggg — CHANGES_REQUESTED** [review]
Will approve once composer release is out, version is bumped, and CI passes

File: scripts/eval/yamls/eval_gauntlet.yaml [hunk 3]
```diff
@@ -115,14 +115,58 @@ eval_gauntlet:
   - name: programming
     benchmarks:
     - name: human_eval
-      num_fewshot: 0
+      num_fewshot: [0, 3]
```
Comment (bmosaicml) [inline scripts/eval/yamls/eval_gauntlet.yaml:118]: should just be "3"

File: scripts/eval/yamls/eval_gauntlet.yaml [hunk 4]
```diff
@@ -115,14 +115,58 @@ eval_gauntlet:
...
       random_baseline: 0.0
     - name: human_eval_cpp
-      num_fewshot: 0
+      num_fewshot: [0, 3]
```
Comment (bmosaicml) [inline scripts/eval/yamls/eval_gauntlet.yaml:121]: same, just "3"

File: scripts/eval/yamls/eval_gauntlet.yaml [hunk 5]
```diff
@@ -115,14 +115,58 @@ eval_gauntlet:
...
       random_baseline: 0.0
     - name: human_eval_js
+      num_fewshot: [0, 3]
```
Comment (bmosaicml) [inline scripts/eval/yamls/eval_gauntlet.yaml:124]: same

**Comment by bmosaicml** [comment]
Lgtm, any chance you could include the time to eval for whichever model/num GPUs you tested this with? It'll be useful reference to know once we run >30B models

File: llmfoundry/utils/builders.py [hunk 6]
```diff
@@ -244,6 +246,10 @@ def _validate_cfg(icl_cfg: DictConfig):
             icl_cfg.max_seq_len = default_max_seq_len
         if 'batch_size' not in icl_cfg:
             icl_cfg.batch_size = default_batch_size
+        if 'pass_at_k' not in icl_cfg:
+            icl_cfg.pass_at_k = 1
+        if 'num_beams' not in icl_cfg:
+            icl_cfg.num_beams = 20
```
[the remark this answers is not in the document]
Comment (mcarbin) [reply llmfoundry/utils/builders.py:252]: this came from a combination of claim in the StarCoder paper and backed by our own analysis/simulation of this choice for pass@1.  Would like a comment to that effect added?

File: scripts/eval/eval.py [hunk 7]
```diff
@@ -91,10 +91,12 @@ def load_model(model_cfg: DictConfig, tokenizer: PreTrainedTokenizerBase,
                     )
 
 
+@torch.no_grad()
```
[the remark this answers is not in the document]
Comment (mcarbin) [reply scripts/eval/eval.py:94]: hmmm..this came from @samhavens. I agree eval should handle this and was unclear on if he was trying to address some other issue.

@samhavens?

File: llmfoundry/utils/builders.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/utils/builders.py:252]: yeah, a comment would be good. but i was also wondering if this choice is really necessary to have defaulted everywhere because I would expect it to slow things down quite a bit

File: scripts/eval/yamls/eval_gauntlet.yaml (hunk 3 above)
Comment (mcarbin) [reply scripts/eval/yamls/eval_gauntlet.yaml:118]: Resolution here is to go with 0 as is consistent with HumanEval methodology in other papers (e.g., CodeLlama, StarCoder, etc)

File: llmfoundry/utils/builders.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcarbin) [reply llmfoundry/utils/builders.py:252]: as a meta question, what is our intention for defaults broadly? Coding guidelines for composer only state they should not be a no-op. 

However, setting num_beams too small produces unreliable results. I would not recommend setting default num_beams = 1 as in the original PR because a user shouldn't trust those results. 

It is also the case that if k is larger, num_beams should perhaps grow (this is open to debate amongst us). So even 20 isn't a uniformly good choice for all k.

Perhaps we shouldn't have a default and instead the user needs to make an explicit choice so they need to think about it?

File: scripts/eval/yamls/eval_gauntlet.yaml (hunk 3 above)
Comment (bmosaicml) [reply scripts/eval/yamls/eval_gauntlet.yaml]: cosigned

File: scripts/eval/yamls/tasks.yaml [hunk 8]
```diff
@@ -176,24 +176,21 @@ icl_tasks:
...
+  num_fewshot: [0, 3]
+  pass_at_k: 1
+  num_beams: 20
   icl_task_type: code_evaluation
 -
   label: human_eval_cpp
   dataset_uri: eval/local_data/programming/processed_human_eval_cpp.jsonl # ADD YOUR OWN DATASET URI
-  num_fewshot: [0]
-  num_beams: 1
+  num_fewshot: [0, 3]
```
Comment (bmosaicml) [inline scripts/eval/yamls/tasks.yaml]: you can make this just [0]

Comment (mcarbin) [reply scripts/eval/yamls/tasks.yaml]: Addressed

File: llmfoundry/utils/builders.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcarbin) [reply llmfoundry/utils/builders.py:252]: Hmm...what do you mean by "it"? leave it as is or remove default?

[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/utils/builders.py:252]: ah sorry, leave it as is.

**Comment by mcarbin** [comment]
For MPT30b, 16xA100-80g is the minimum spec. Here are reference results:
```
Ran mosaicml/mpt-30b eval in: 15220.648153305054 seconds 
Printing complete results for all models 
| Category   | Benchmark      | Subtask   |   Accuracy | Number few shot   | Model            | 
|:-----------|:---------------|:----------|-----------:|:------------------|:-----------------| 
|            | human_eval     |           |  0.143598  | 0-shot            | mosaicml/mpt-30b | 
|            | human_eval_cpp |           |  0.0928571 | 0-shot            | mosaicml/mpt-30b | 
|            | human_eval_js  |           |  0.0951219 | 0-shot            | mosaicml/mpt-30b | 
|            | human_eval_c   |           |  0.0888889 | 0-shot            | mosaicml/mpt-30b | 
```

File: scripts/eval/eval.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mcarbin) [reply scripts/eval/eval.py]: went ahead and removed this.

[the remark this answers is not in the document]
Comment (samhavens) [reply scripts/eval/eval.py]: Thank you. Was checking to see if turning on no grad sooner helped with the issue we were having where FSDP init was hanging in torch 2 (it did not)

File: llmfoundry/utils/builders.py [hunk 9]
```diff
@@ -244,6 +246,10 @@ def _validate_cfg(icl_cfg: DictConfig):
             icl_cfg.max_seq_len = default_max_seq_len
         if 'batch_size' not in icl_cfg:
             icl_cfg.batch_size = default_batch_size
+        if 'pass_at_k' not in icl_cfg:
```
Comment (dakinggg) [inline llmfoundry/utils/builders.py:249]: I think these should only be added if the task is code eval, right?

**Review by dakinggg — APPROVED** [review]
